<!-- 
  Simulate a differential drive robot with the mpc_local_planner in stage:
  - stage
  - map_server
  - move_base
  - static map
  - amcl
  - rviz view
 -->
<launch>
	<arg name="enable_vis" default="$(optenv GM_PARAM_ENABLE_VIS false)"/>

	<param name="ego_circle_topic" value="/point_scan"/>

	<node name="egocircle_node" pkg="egocircle" type="ego_circle">
    </node>
        <!--  ************** Navigation ***************  -->

	<arg name="far_feasible" default="$(optenv GM_PARAM_FAR_FEASIBLE true)"/>

    <include file="$(find turtlebot_navigation)/launch/includes/velocity_smoother.launch.xml"/>

	<arg name="v_des" default="$(optenv GM_PARAM_V_DES 0.3)"/>
	<arg name="use_bezier" default="$(optenv GM_PARAM_USE_BEZIER true)"/>
	<arg name="use_pose_controller" default="$(optenv GM_PARAM_USE_POSE_CONTROLLER false)"/>
	<arg name="use_keyhole" default="$(optenv GM_PARAM_USE_KEYHOLE true)"/>
	<arg name="use_po" default="$(optenv GM_PARAM_USE_PO false)"/>
	<arg name="use_ni" default="$(optenv GM_PARAM_USE_NI true)"/>
	<arg name="has_feedforward" default="$(optenv GM_PARAM_HAS_FEEDFORWARD true)"/>
    
	<node pkg="move_base_virtual" type="move_base_virtual" respawn="false" name="move_base" output="screen">
  	  	<rosparam file="$(find safer_gap_benchmark)/config/common_costmap_params.yaml" command="load" ns="global_costmap"/>
        <rosparam file="$(find safer_gap_benchmark)/config/common_costmap_params.yaml" command="load" ns="local_costmap"/>
        <rosparam file="$(find safer_gap_benchmark)/config/local_costmap_params.yaml" command="load"/>
        <rosparam file="$(find safer_gap_benchmark)/config/global_costmap_params.yaml" command="load"/>
  		<rosparam file="$(find safer_gap_benchmark)/config/pg_mpc_local_planner_params_casadi.yaml" command="load" />

        <rosparam file="$(find safer_gap_benchmark)/config/move_base_params.yaml" command="load"/>
        <rosparam file="$(find safer_gap_benchmark)/config/global_planner_params.yaml" command="load" ns="GlobalPlanner"/>

		<param name="planner_frequency" value="1.0" />
		<param name="planner_patience" value="5.0" />

		<param name="base_local_planner" value="sg_mpc_local_planner/SGMpcHcLocalPlannerROS" />
		<param name="controller_frequency" value="5.0" />
		<param name="controller_patience" value="15.0" />

        <remap from="cmd_vel" to="navigation_velocity_smoother/raw_cmd_vel"/>

		<rosparam file="$(find safer_gap_benchmark)/config/potentialgap_local_planner_params.yaml" command="load" ns="SGMpcHcLocalPlannerROS"/>

		<param name="SGMpcHcLocalPlannerROS/shape_id" value="0"/>
		<param name="SGMpcHcLocalPlannerROS/length" value="0.4"/>

		<param name="SGMpcHcLocalPlannerROS/far_feasible" value="$(arg far_feasible)"/>
		<param name="SGMpcHcLocalPlannerROS/pf_local_frame_enable" value="false"/>

		<param name="SGMpcHcLocalPlannerROS/use_bezier" value="$(arg use_bezier)"/>
		<param name="SGMpcHcLocalPlannerROS/radial_extend" value="false" if="$(arg use_bezier)"/>
		<param name="SGMpcHcLocalPlannerROS/radial_extend" value="true" unless="$(arg use_bezier)"/>
		<param name="SGMpcHcLocalPlannerROS/inflated_gap" value="true" if="$(arg use_keyhole)"/>
		<param name="SGMpcHcLocalPlannerROS/inflated_gap" value="false" unless="$(arg use_keyhole)"/>

		<param name="SGMpcHcLocalPlannerROS/map_frame_id" value="map"/>
		<param name="SGMpcHcLocalPlannerROS/odom_frame_id" value="odom"/>
		<param name="SGMpcHcLocalPlannerROS/robot_frame_id" value="base_footprint"/>
		<param name="SGMpcHcLocalPlannerROS/sensor_frame_id" value="camera_depth_frame"/>
		<param name="SGMpcHcLocalPlannerROS/base_frame_id" value="base_footprint"/>

		<param name="SGMpcHcLocalPlannerROS/robot_path_orient_linear_decay" value="false"/>
		<param name="SGMpcHcLocalPlannerROS/virtual_path_decay_enable" value="false"/>
		<param name="SGMpcHcLocalPlannerROS/inflated_gap_scale" value="1.1"/> <!--1.-->

		<param name="SGMpcHcLocalPlannerROS/print/print_debug_info" value="false"/>
		<param name="SGMpcHcLocalPlannerROS/print/print_timing" value="true"/>

		<param name="SGMpcHcLocalPlannerROS/solver/linear_solver" value="ma27"/> <!-- Available solvers: ma27, ma57, ma77, ma86, ma97, pardiso, wsmp, mumps -->
		<param name="SGMpcHcLocalPlannerROS/solver/N" value="6"/>
		<param name="SGMpcHcLocalPlannerROS/robot/v_max" value="0.5"/>
		<param name="SGMpcHcLocalPlannerROS/robot/w_min" value="-4"/>
		<param name="SGMpcHcLocalPlannerROS/robot/w_max" value="4"/>
		<param name="SGMpcHcLocalPlannerROS/robot/v_a_max" value="10"/> <!--0.55-->
		<param name="SGMpcHcLocalPlannerROS/robot/w_a_max" value="1.78"/> <!--1.78-->
		<param name="SGMpcHcLocalPlannerROS/solver/Q1" value="10"/>
		<param name="SGMpcHcLocalPlannerROS/solver/Q2" value="10"/>
		<param name="SGMpcHcLocalPlannerROS/solver/Q3" value="0"/> <!--theta has problem, need to set to 0-->
		<param name="SGMpcHcLocalPlannerROS/solver/R1" value="1"/>
		<param name="SGMpcHcLocalPlannerROS/solver/R2" value="1"/>
		<param name="SGMpcHcLocalPlannerROS/solver/terminal_weight" value="20"/>
		<param name="SGMpcHcLocalPlannerROS/solver/orient_weight" value="20"/>

		<param name="SGMpcHcLocalPlannerROS/controller/param/v_des" value="$(arg v_des)"/> <!-- 0.3 0.4 -->
		<param name="SGMpcHcLocalPlannerROS/controller/param/a_des" value="1"/>

		<param name="SGMpcHcLocalPlannerROS/controller/use_pose_controller" value="$(arg use_pose_controller)"/>
		<param name="SGMpcHcLocalPlannerROS/controller/has_feedforward" value="$(arg has_feedforward)"/>

		<param name="SGMpcHcLocalPlannerROS/controller/pose_con/k_drive_x" value="1" if="$(arg has_feedforward)"/>
		<param name="SGMpcHcLocalPlannerROS/controller/pose_con/k_drive_y" value="1" if="$(arg has_feedforward)"/>
		<param name="SGMpcHcLocalPlannerROS/controller/pose_con/k_turn" value="1" if="$(arg has_feedforward)"/>
		<param name="SGMpcHcLocalPlannerROS/controller/pose_con/k_drive_x" value="2" unless="$(arg has_feedforward)"/> <!--2-->
		<param name="SGMpcHcLocalPlannerROS/controller/pose_con/k_drive_y" value="3.5" unless="$(arg has_feedforward)"/> <!--3.5-->
		<param name="SGMpcHcLocalPlannerROS/controller/pose_con/k_turn" value="0.5" unless="$(arg has_feedforward)"/> <!--0.5-->

		<param name="SGMpcHcLocalPlannerROS/controller/use_keyhole" value="$(arg use_keyhole)"/>
		<param name="SGMpcHcLocalPlannerROS/controller/use_cbf" value="false"/>

		<param name="SGMpcHcLocalPlannerROS/controller/use_po" value="$(arg use_po)"/>
		<param name="SGMpcHcLocalPlannerROS/controller/po/k_po" value="1.2"/> <!--1.2-->
		<param name="SGMpcHcLocalPlannerROS/controller/po/k_po_turn" value="1"/> <!--1-->
		<param name="SGMpcHcLocalPlannerROS/controller/po/r_min" value="0.36"/> <!--0.36-->
		<param name="SGMpcHcLocalPlannerROS/controller/po/r_inscr" value="0.2"/> <!--0.2-->

		<param name="SGMpcHcLocalPlannerROS/cbf/cbf_gamma" value="1"/>

		<param name="SGMpcHcLocalPlannerROS/controller/ni/ni_enabled" value="$(arg use_ni)"/>
		<param name="NIConfig/v_max" value="0.5" /> <!--0.7 0.5-->
		<param name="NIConfig/w_max" value="4" /> <!--1 4-->
		<param name="NIConfig/a_max" value="0.55" /> <!--0.2 0.55-->
		<param name="NIConfig/w_dot_max" value="1.78" /> <!--0.8 1.78-->
		<param name="NIConfig/c_p" type="double" value="1"/>
		<param name="NIConfig/c_d" type="double" value="5"/>
		<param name="NIConfig/c_lambda" type="double" value="1"/>
		<param name="NIConfig/epsilon" type="double" value="0.1"/>

		<param name="SGMpcHcLocalPlannerROS/viz/enable_vis" value="$(arg enable_vis)"/>

	</node>

</launch>
